returnn/hdf.py

```diff
@@ -346,6 +364,296 @@ def run(self):
+        _logging.info(f"Writing audio to {out_hdfs} from {splits}.")
+
+        SimpleHDFWriter = get_returnn_simple_hdf_writer(self.returnn_root.get_path())
+        hdf_writers = [SimpleHDFWriter(filename=out_hdf.get_path(), dim=1) for out_hdf in out_hdfs]
+        segment_whitelists = []
+        for split in splits:
+            with uopen(split, "rt") as f:
+                segments_whitelist = {line.strip() for line in f if len(line.strip()) > 0}
+            segment_whitelists.append(segments_whitelist)
+
+        assert len(segment_whitelists) == len(out_hdfs)
+        all_whitelists = set.union(*segment_whitelists)
+
+        c = corpus.Corpus()
+        c.load(self.bliss_corpus.get_path())
+
+        pool = multiprocessing.Pool(self.num_workers)
+        recs = (
+            # We send every segment of the recording grouped together with the audio to
+            # avoid having to read the audio file more than once.
+            #
+            # We only pass down the required audio/segment metadata instead of the whole
+            # `i6_core.lib.corpus.Segment` object to avoid excessive pickling overhead,
+            # stalling the worker processes (if we don't).
+            (rec.audio, segments)
+            for rec in c.all_recordings()
+            if len(
+                segments := [
+                    (full_name, segment.start, segment.end)
+                    for segment in rec.segments
+                    if (full_name := segment.fullname()) in all_whitelists
+                ]
+            )
+            > 0
+        )
+        # use imap instead of imap_unordered to have reproducible results, even though this
+        # causes a slight speed penalty
+        for results in pool.imap(self._process_seq, recs, chunksize=8):
+            for (segment_name, data), (hdf_writer, segments_whitelist) in itertools.product(
+                results, zip(hdf_writers, segment_whitelists)
+            ):
+                if segment_name not in segments_whitelist:
+                    continue
+                _logging.info(f"Writing {segment_name} to {hdf_writer.filename}.")
+                hdf_writer.insert_batch(
+                    inputs=data.reshape(1, -1, 1),
+                    seq_len=[data.shape[0]],
+                    seq_tag=[segment_name],
+                )
+        for hdf_writer in hdf_writers:
+            hdf_writer.close()
+
+    def _process_seq(self, recording: Tuple[str, Sequence[Tuple[str, float, float]]]) -> List[Tuple[str, np.ndarray]]:
+        audio_file, segments = recording
+
+        assert isinstance(self.multi_channel_strategy, (BlissToPcmHDFJob.Mixdown, BlissToPcmHDFJob.PickNth)), (
+            "unknown multi_channel_strategy"
+        )
+        channel_mix_args = (
+            ["-ac", "1"]
+            if isinstance(self.multi_channel_strategy, BlissToPcmHDFJob.Mixdown)
+            else ["-map_channel", f"0.{self.multi_channel_strategy.channel}"]
+        )
```

Comment:
Btw, I wonder a bit: You expect that the user of `BlissToAudioHDFJob` always wants to get a single audio channel out? But at the same time, you allow to write any custom audio format into the HDF. Doesn't it make sense to also allow for multiple channels?

I can see two further reasonable cases here:
* `multi_channel_strategy is None` -> `channel_mix_args = []`, i.e. do nothing, leave it as it is
* `multi_channel_strategy` is sth like `CustomFfmpegChannelArgs` (or so, maybe better name) -> `channel_mix_args = multi_channel_strategy.channel_mix_args`, i.e. just specify any custom args

Reply:
Yeah, this is mainly because the other job doesn't allow for multiple channels either at the moment. I agree, we can in any case be more relaxed about this in the future :+1: